CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed

marta- opened the pull request:
**Includes:**
* **[CARDS-683](https://phenotips.atlassian.net/browse/CARDS-683) - _As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed_**
* [CARDS-1764](https://phenotips.atlassian.net/browse/CARDS-1764) - _Questionnaire wizard: the variable name / entry identifiers should be visible in the edit mode_
* [CARDS-1765](https://phenotips.atlassian.net/browse/CARDS-1765) - _Better display of the `expression` property of computed fields, as a formatted code block_
* [CARDS-1766](https://phenotips.atlassian.net/browse/CARDS-1766) - _When deleting an item, reload only its parent instead of the whole questionnaire_
* Some refactoring and console warning fixes

**To test:**
* The entire functionality of the questionnaire wizard in general
* Creating / updating section conditions in the questionnaire wizard
* In forms, the behavior of sections with conditions created/updated in the questionnaire wizard

**Reviewing:**
* This is a large PR that adresses several related issues and affecting multiple files. A couple of refactoring iterations were done prior to the implementation of the actual new feature. Reviewing **commit by commit** rather than file by file is recommended.

marta- committed:
CARDS-683 - As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Refactor the questionnaire wizard components for questionnaire entries to reduce duplicate logic

marta- committed:
CARDS-683 - As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Refactor how the values of properties of questionnaire items are displayed in the questionnaire wizard, using a component manager

marta- committed:
CARDS-1765 - Better display of the `expression` property of computed fields, as a formatted code block

marta- committed:
CARDS-1766 - When deleting an item, reload only its parent instead of the whole questionnaire

marta- committed:
[misc] More expressive names for handler functions passed between components in the questionnaire wizard

marta- committed:
CARDS-1764 - Questionnaire wizard: the variable name / entry identifiers should be visible in the edit mode
Additionally, the field being used for the Questionnaire entry title should not be displayed again in the field list
(e.g. `text` for questions, `label` for sections).

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
* Added the Conditional and ConditionalGroup entry types
* Added support for entry grouping (e.g. Condition = { ConditionalGroup, Conditional}) and max limit to the number of children of a type inside a parent (to limit to max 1 Conditional OR ConditionalGroup per Section)
* Added support for entry priorities (to display conditions at the top of a section regardless where they are in the Questionnaire definition)

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
* Use a dropdown for variable names
* Display the question text of the conditionalValue is a reference

marta- committed:
[misc] Fix console warnings when creating a new item in the questionnaire editor

marta- committed:
[misc] Preserve the order of children inside a cards:ConditionalGroup and cards:Conditional

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Use consistent terminology: "Question id" (just like "Section id", ...) instead of "Variable name"

marta- committed:
CARDS-1764 - Questionnaire wizard: the variable name / entry identifiers should be visible in the edit mode
Simpler display of the entry id

marta- committed:
CARDS-1764 - Questionnaire wizard: the variable name / entry identifiers should be visible in the edit mode
Simplify the markup for displaying type, id and title

marta- committed:
[misc] Missing ":" after id label in the questionnaire wizard's edit dialog

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Added support for unary operators "is empty" and "is not empty"

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -239,23 +239,99 @@ export default withStyles(QuestionnaireStyle)(Questionnaire);
+    //   f: "f.json",
+    //   group3 : {entries: {g: "g,json", h: "h.json"}, defaultOrder: 1}
+    // }
+    // => {g: "g.json", h: "h,json", h: "h.json", e: "e.json"}
```
```suggestion
    // => {g: "g.json", h: "h,json", d: "d.json", e: "e.json"}
```

acrowthe commented:
Bug: In the admin questionnaire editor, adding new items with the blue (+) floating action button correctly adds the item but does not update the view to display them, requiring a refresh to view the newly added item.

Bug found, but fixed by rebasing onto dev: Conditional ranges aren't working properly, should work fine once merged

marta- committed:
[misc] More expressive names for handler functions passed between components in the questionnaire wizard
Bug fix: One forgotten param name change caused the questionnaire not to display newly added direct children

acrowthe reviewed (approved):
Looks good once tested to confirm the (+) fab updates the list now

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -239,23 +239,99 @@ export default withStyles(QuestionnaireStyle)(Questionnaire);
+      if ( typeof(v) == "object" && typeof(v?.entries) != "undefined" && typeof(v?.defaultOrder) == "undefined") {
+        // Flatten groups with `entries` but without `defaultOrder` (the ones with defaultOrder are already in the "priority" list)
+        generalModels = {...generalModels, ...v.entries}
+      } else {
```
This `else` will match all failures: v is not an object, v is an object but does not have entries, v is an object with entries and a default order. Based on the comment above, I think you want just the case when v is a simple string, right?

marta- replied:
Addressed in 18a5aec.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -239,23 +239,99 @@ export default withStyles(QuestionnaireStyle)(Questionnaire);
+    });
+    // If there are any entries without defaultOrder, we update the generalEntryTypes
+    // Otherwise the original list is kept, i.e. ENTRY_TYPES
+    if (generalModels != {}) {
```
Uh, I don't like this check. It happens that `{} != {}`.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -239,23 +239,99 @@ export default withStyles(QuestionnaireStyle)(Questionnaire);
+                         classes={classes}
+                       />
+                     </Grid>
+        )(eval(_stripCardsNamespace(value['jcr:primaryType'])))
```
Took me a while to figure out what's going on here. Could use a comment.

I'm not fond of evals, can there be a map between name and component?

marta- replied:
Not in this iteration. This is not new code.

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed

`QuestionnaireItemSet` child entry ordering:
* Fixed comment detailing a questionnaire entry ordering example
* Fixed wrong way to check if an object is empty

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed

Added a new comment to a function encapsulating old code, for clarification.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -239,23 +239,117 @@ export default withStyles(QuestionnaireStyle)(Questionnaire);
+  //     To call the right component for each entry that has passed the "types" filter,  we strip its primaryType
+  //     of the "cards:" prefix and then `eval` the result to the functional component's name, which is passed as
+  //     a parameter to an anonymous function called for each such entry, that renders the component inside a
+  //     Grid itme.
```
```suggestion
  //     Grid item.
```

marta- replied:
Fixed in d9d43d2.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ConditionalValueInput.jsx:
```diff
@@ -0,0 +1,339 @@
+    if (event?.relatedTarget?.type == "submit") {
+      const timer = setTimeout(() => {
+        saveButtonRef?.current?.click();
+      }, 500);
```
That seems fragile, if the OS is busy, it may happen that the needed operation takes longer than 500ms. But then this is something that an admin would do, so not likely to affect users.

marta- replied:
This is not new code, it's being adapted from AnswerOptions. Please make a task to refactor both.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ConditionalValueInput.jsx:
```diff
@@ -0,0 +1,339 @@
+        })
+    }
+  }, []);
+
```
I wonder if it's better to put `parentQuestionnaire` as a dependency here.

marta- replied:
Won't make a difference.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ConditionalValueInput.jsx:
```diff
@@ -0,0 +1,339 @@
+    setVariables(vars);
+  }
+
+  let findQuestions = (json, result) =>  {
```
This and the one above seem identical to the edit component. Any way to have only one definition?

marta- replied:
I had attempted a single component for both view and edit, ended up with code that was harder to read, a few ifs and a couple of bugs.

Can we leave the refactoring for later?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ConditionalValueInput.jsx:
```diff
@@ -0,0 +1,339 @@
+    })
+  }
+
+   return (
```
The indentation seems off, is it?

marta- replied:
Fixed in b47f2aa.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ConditionalValueInput.jsx:
```diff
@@ -0,0 +1,339 @@
+export default ConditionalValueInput;
+
+QuestionComponentManager.registerQuestionComponent((definition) => {
+  if (["conditionalValue"].includes(definition)) {
```
Strange way to check if two strings are equal.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/MarkdownTextField.jsx:
```diff
@@ -53,3 +55,15 @@ QuestionComponentManager.registerQuestionComponent((definition) => {
+    if (definition == "markdown") {
+      return [MarkdownTextValue, 50];
+    }
+});
```
Add newline at the end of file.

marta- replied:
Done in b47f2aa.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ValueComponentManager.jsx:
```diff
@@ -0,0 +1,42 @@
+  // Registers a new value displayer. This method is only supposed to be called by each value displayer.
+  //
+  // @param component A function which can be called to determine how well suited the registered component can display a specific question.
+  //     The function receives the JSON of a question definition, and returns a tuple (array with 2 items), the actual React Component that can display the question, and the confidence that it is the right displayer.
```
These 2 lines seem wrong, as it's not displaying questions, and as I can see it receives a simple string, not a JSON definition of a question.

marta- replied:
Fixed in d9d43d2.

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Addressing review comments: Comment fixes

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Addressing review comments: code style fixes

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Addressing review comments: fixed the condition for displaying listing non-prioritary entry types

sashaandjic reviewed (approved):
tested

marta- committed:
CARDS-683: As an admin I can use the questionnaire wizard to specify the condition or conditions when a questionnaire section should be displayed
Updated the new component to mui5+

marta- commented:
Rebased on the latest `dev` which includes Material UI 5 upgrade, and updated the new component to Material UI 5 in f99e55a.

sashaandjic reviewed (approved):
tested
